Tried to send an out-of-range integer as a 2-byte value when deleting large model

# Description

Deleting a large data model causes a system exception and the database connection to be closed

# Steps to reproduce
Take a large data model, permanently delete it.  It runs for a second, and then dies.  Further requests all result in an exception because the database connection has been closed.

# Expected behavior

Not that.

# Error details

```
avax.persistence.PersistenceException: org.hibernate.exception.JDBCConnectionException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.query.internal.AbstractProducedQuery.executeUpdate(AbstractProducedQuery.java:1705)
	at javax.persistence.Query$executeUpdate.call(Unknown Source)
	at uk.ac.ox.softeng.maurodatamapper.core.traits.service.MultiFacetAwareService$Trait$Helper.deleteAllFacetsByMultiFacetAwareIds(MultiFacetAwareService.groovy:173)
	at uk.ac.ox.softeng.maurodatamapper.core.traits.service.MultiFacetAwareService$Trait$Helper$deleteAllFacetsByMultiFacetAwareIds.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at uk.ac.ox.softeng.maurodatamapper.core.traits.service.MultiFacetAwareService$Trait$Helper$deleteAllFacetsByMultiFacetAwareIds.call(Unknown Source)
	at uk.ac.ox.softeng.maurodatamapper.core.model.CatalogueItemService.deleteAllFacetsByMultiFacetAwareIds(CatalogueItemService.groovy)
	at uk.ac.ox.softeng.maurodatamapper.core.traits.service.MultiFacetAwareService$deleteAllFacetsByMultiFacetAwareIds$2.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:194)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.item.datatype.enumeration.EnumerationValueService.deleteAllByModelIds(EnumerationValueService.groovy:90)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.item.datatype.enumeration.EnumerationValueService$deleteAllByModelIds.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.item.datatype.DataTypeService.$tt__dataTypeService_deleteAllByModelIds(DataTypeService.groovy:128)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1268)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1029)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1012)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:101)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.item.datatype.DataTypeService$_deleteAllByModelIds_closure8.doCall(DataTypeService.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at groovy.lang.Closure.call(Closure.java:428)
	at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.item.datatype.DataTypeService.deleteAllByModelIds(DataTypeService.groovy)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.item.datatype.DataTypeService$deleteAllByModelIds$0.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService.$tt__dataModelService_deleteModelsAndContent(DataModelService.groovy:478)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1268)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1029)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1012)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:101)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService$_deleteModelsAndContent_closure20.doCall(DataModelService.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at groovy.lang.Closure.call(Closure.java:428)
	at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService.deleteModelsAndContent(DataModelService.groovy)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService$deleteModelsAndContent$1.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at uk.ac.ox.softeng.maurodatamapper.core.model.ModelService.deleteModelAndContent(ModelService.groovy:196)
	at uk.ac.ox.softeng.maurodatamapper.core.model.ModelService$deleteModelAndContent$4.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService.$tt__dataModelService_delete(DataModelService.groovy:250)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1268)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1029)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1012)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:101)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService$_delete_closure13.doCall(DataModelService.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at groovy.lang.Closure.call(Closure.java:428)
	at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService.delete(DataModelService.groovy)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService.delete(DataModelService.groovy)
	at uk.ac.ox.softeng.maurodatamapper.datamodel.DataModelService$delete$0.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:194)
	at uk.ac.ox.softeng.maurodatamapper.core.model.ModelService.permanentDeleteModel(ModelService.groovy:222)
	at uk.ac.ox.softeng.maurodatamapper.core.model.ModelService$permanentDeleteModel$3.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
	at uk.ac.ox.softeng.maurodatamapper.core.controller.ModelController.$tt__modelController_delete(ModelController.groovy:153)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1268)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1029)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1012)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:101)
	at uk.ac.ox.softeng.maurodatamapper.core.controller.ModelController$_delete_closure2.doCall(ModelController.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at groovy.lang.Closure.call(Closure.java:428)
	at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at uk.ac.ox.softeng.maurodatamapper.core.controller.ModelController.delete(ModelController.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doDelete(FrameworkServlet.java:931)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:702)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:96)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:91)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
```

```
Caused by: java.io.IOException: Tried to send an out-of-range integer as a 2-byte value: 227925
	at org.postgresql.core.PGStream.sendInteger2(PGStream.java:359)
	at org.postgresql.core.v3.QueryExecutorImpl.sendParse(QueryExecutorImpl.java:1604)
	at org.postgresql.core.v3.QueryExecutorImpl.sendOneQuery(QueryExecutorImpl.java:1929)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1487)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:347)
	... 216 common frames omitted
```
# Environment

Running the latest versions of core and m-a-b.

# Additional context

Seems to be another occurrence of #354 but occurring in a different part of the code.

--- a/mdm-common/src/main/groovy/uk/ac/ox/softeng/maurodatamapper/util/Utils.groovy
+++ b/mdm-common/src/main/groovy/uk/ac/ox/softeng/maurodatamapper/util/Utils.groovy
@@ -22,6 +22,8 @@ import grails.config.Config
 import grails.core.GrailsApplication
 import grails.core.GrailsClass
 import groovy.transform.CompileStatic
+import groovy.transform.stc.ClosureParams
+import groovy.transform.stc.SimpleType
 import groovy.util.logging.Slf4j
 import org.grails.core.artefact.DomainClassArtefactHandler
 import org.slf4j.Logger
@@ -226,5 +228,20 @@ class Utils {
         }
         return partitions
     }
+
+    public final static Integer BATCH_SIZE = 10000
+
+/*
+    This method executes a closure in batches to ensure that we don't inadvertently call with too many ids and
+    break Postgres: Tried to send an out-of-range integer as a 2-byte value
+ */
+
+    static void executeInBatches(List items, @ClosureParams(value= SimpleType.class, options="java.util.List") Closure closure) {
+        List<List> partitions = partition(items, BATCH_SIZE)
+        partitions.each { partition ->
+            closure.call(partition)
+        }
+    }
+
 
 }

--- a/mdm-core/grails-app/services/uk/ac/ox/softeng/maurodatamapper/core/facet/BreadcrumbTreeService.groovy
+++ b/mdm-core/grails-app/services/uk/ac/ox/softeng/maurodatamapper/core/facet/BreadcrumbTreeService.groovy
@@ -56,12 +56,13 @@ class BreadcrumbTreeService {
     void deleteAllByDomainIds(Set<UUID> domainIds) {
         List<String> idPrefixesToDelete = domainIds.collect {it.toString() + '|'}
 
-        sessionFactory.currentSession
-            .createSQLQuery('DELETE FROM core.breadcrumb_tree WHERE SUBSTR(tree_string, 1, :id_length+1) IN :id_prefixes')
-            .setParameter('id_length', Utils.UUID_CHARACTER_LENGTH)
-            .setParameterList('id_prefixes', idPrefixesToDelete)
-            .executeUpdate()
-
+        Utils.executeInBatches(idPrefixesToDelete, { ids ->
+            sessionFactory.currentSession
+                    .createSQLQuery('DELETE FROM core.breadcrumb_tree WHERE SUBSTR(tree_string, 1, :id_length+1) IN :id_prefixes')
+                    .setParameter('id_length', Utils.UUID_CHARACTER_LENGTH)
+                    .setParameterList('id_prefixes', ids)
+                    .executeUpdate()
+        })
         log.trace('BreadcrumbTrees removed')
     }
 }

--- a/mdm-core/src/main/groovy/uk/ac/ox/softeng/maurodatamapper/core/traits/service/MultiFacetAwareService.groovy
+++ b/mdm-core/src/main/groovy/uk/ac/ox/softeng/maurodatamapper/core/traits/service/MultiFacetAwareService.groovy
@@ -168,12 +168,15 @@ trait MultiFacetAwareService<K extends MultiFacetAware> {
     }
 
     void deleteAllFacetsByMultiFacetAwareIds(List<UUID> multiFacetAwareIds, String queryToDeleteFromJoinTable) {
-        sessionFactory.currentSession
-            .createSQLQuery(queryToDeleteFromJoinTable)
-            .setParameter('ids', multiFacetAwareIds)
-            .executeUpdate()
 
-        deleteAllFacetDataByMultiFacetAwareIds multiFacetAwareIds
+        Utils.executeInBatches(multiFacetAwareIds, {ids ->
+            sessionFactory.currentSession
+                .createSQLQuery(queryToDeleteFromJoinTable)
+                .setParameter('ids', ids)
+                .executeUpdate()
+
+            deleteAllFacetDataByMultiFacetAwareIds ids
+        })
     }
 
     void deleteAllFacetDataByMultiFacetAwareIds(List<UUID> multiFacetAwareIds) {

--- a/mdm-plugin-dataflow/grails-app/services/uk/ac/ox/softeng/maurodatamapper/dataflow/DataFlowService.groovy
+++ b/mdm-plugin-dataflow/grails-app/services/uk/ac/ox/softeng/maurodatamapper/dataflow/DataFlowService.groovy
@@ -105,11 +105,13 @@ class DataFlowService extends ModelItemService<DataFlow> {
                                                 'delete from dataflow.join_dataflow_to_facet where dataflow_id in :ids')
 
             log.trace('Removing {} DataFlows', dataFlowIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('DELETE FROM dataflow.data_flow WHERE source_id in :ids OR target_id in :ids')
-                .setParameter('ids', modelIds)
-                .executeUpdate()
 
+            Utils.executeInBatches(modelIds, { ids ->
+                sessionFactory.currentSession
+                        .createSQLQuery('DELETE FROM dataflow.data_flow WHERE source_id in :ids OR target_id in :ids')
+                        .setParameter('ids', ids)
+                        .executeUpdate()
+            })
             log.trace('DataFlows removed')
         }
     }

--- a/mdm-plugin-dataflow/grails-app/services/uk/ac/ox/softeng/maurodatamapper/dataflow/component/DataClassComponentService.groovy
+++ b/mdm-plugin-dataflow/grails-app/services/uk/ac/ox/softeng/maurodatamapper/dataflow/component/DataClassComponentService.groovy
@@ -113,25 +113,30 @@ class DataClassComponentService extends ModelItemService<DataClassComponent> {
             dataElementComponentService.deleteAllByModelIds(modelIds)
 
             log.trace('Removing links to DataClasses in {} DataClassComponents', dataClassComponentIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('delete from dataflow.join_data_class_component_to_source_data_class where data_class_component_id in :ids')
-                .setParameter('ids', dataClassComponentIds)
-                .executeUpdate()
+            Utils.executeInBatches(dataClassComponentIds, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('delete from dataflow.join_data_class_component_to_source_data_class where data_class_component_id in :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
 
-            sessionFactory.currentSession
-                .createSQLQuery('delete from dataflow.join_data_class_component_to_target_data_class where data_class_component_id in :ids')
-                .setParameter('ids', dataClassComponentIds)
-                .executeUpdate()
+                sessionFactory.currentSession
+                    .createSQLQuery('delete from dataflow.join_data_class_component_to_target_data_class where data_class_component_id in :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+            })
 
             log.trace('Removing facets for {} DataClassComponents', dataClassComponentIds.size())
             deleteAllFacetsByMultiFacetAwareIds(dataClassComponentIds,
                                                 'delete from dataflow.join_dataclasscomponent_to_facet where dataclasscomponent_id in :ids')
 
             log.trace('Removing {} DataClassComponents', dataClassComponentIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('delete from dataflow.data_class_component where id in :ids')
-                .setParameter('ids', dataClassComponentIds)
-                .executeUpdate()
+
+            Utils.executeInBatches(dataClassComponentIds, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('delete from dataflow.data_class_component where id in :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+            })
 
             log.trace('DataClassComponents removed')
         }

--- a/mdm-plugin-dataflow/grails-app/services/uk/ac/ox/softeng/maurodatamapper/dataflow/component/DataElementComponentService.groovy
+++ b/mdm-plugin-dataflow/grails-app/services/uk/ac/ox/softeng/maurodatamapper/dataflow/component/DataElementComponentService.groovy
@@ -106,25 +106,29 @@ class DataElementComponentService extends ModelItemService<DataElementComponent>
         if (dataElementComponentIds) {
 
             log.trace('Removing links to DataElements in {} DataElementComponents', dataElementComponentIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('delete from dataflow.join_data_element_component_to_source_data_element where data_element_component_id in :ids')
-                .setParameter('ids', dataElementComponentIds)
-                .executeUpdate()
+            Utils.executeInBatches(dataElementComponentIds, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('delete from dataflow.join_data_element_component_to_source_data_element where data_element_component_id in :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
 
-            sessionFactory.currentSession
-                .createSQLQuery('delete from dataflow.join_data_element_component_to_target_data_element where data_element_component_id in :ids')
-                .setParameter('ids', dataElementComponentIds)
-                .executeUpdate()
+                sessionFactory.currentSession
+                    .createSQLQuery('delete from dataflow.join_data_element_component_to_target_data_element where data_element_component_id in :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+            })
 
             log.trace('Removing facets for {} DataElementComponents', dataElementComponentIds.size())
             deleteAllFacetsByMultiFacetAwareIds(dataElementComponentIds,
                                                 'delete from dataflow.join_dataelementcomponent_to_facet where dataelementcomponent_id in :ids')
 
             log.trace('Removing {} DataElementComponents', dataElementComponentIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('delete from dataflow.data_element_component where id in :ids')
-                .setParameter('ids', dataElementComponentIds)
-                .executeUpdate()
+            Utils.executeInBatches(dataElementComponentIds, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('delete from dataflow.data_element_component where id in :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+            })
 
             log.trace('DataElementComponents removed')
         }

--- a/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/DataModelService.groovy
+++ b/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/DataModelService.groovy
@@ -481,10 +481,12 @@ class DataModelService extends ModelService<DataModel> implements SummaryMetadat
         deleteAllFacetsByMultiFacetAwareIds(idsToDelete.toList(), 'delete from datamodel.join_datamodel_to_facet where datamodel_id in :ids')
 
         log.trace('Content removed')
-        sessionFactory.currentSession
-            .createSQLQuery('DELETE FROM datamodel.data_model WHERE id IN :ids')
-            .setParameter('ids', idsToDelete)
-            .executeUpdate()
+        Utils.executeInBatches(idsToDelete as List, {ids ->
+            sessionFactory.currentSession
+                .createSQLQuery('DELETE FROM datamodel.data_model WHERE id IN :ids')
+                .setParameter('ids', ids)
+                .executeUpdate()
+        })
 
         log.trace('DataModels removed')
 

--- a/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataClassService.groovy
+++ b/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataClassService.groovy
@@ -215,10 +215,13 @@ class DataClassService extends ModelItemService<DataClass> implements SummaryMet
                                                 'delete from datamodel.join_dataclass_to_facet where dataclass_id in :ids')
 
             log.trace('Removing {} DataClasses', dataClassIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('DELETE FROM datamodel.data_class WHERE data_model_id IN :ids')
-                .setParameter('ids', dataModelIds)
-                .executeUpdate()
+
+            Utils.executeInBatches(dataModelIds as List, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('DELETE FROM datamodel.data_class WHERE data_model_id IN :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+            })
 
             log.trace('DataClasses removed')
         }

--- a/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataElementService.groovy
+++ b/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataElementService.groovy
@@ -117,11 +117,11 @@ class DataElementService extends ModelItemService<DataElement> implements Summar
 
         if (dataElementIds) {
             log.trace('Removing facets for {} DataElements', ids.size())
-            deleteAllFacetsByMultiFacetAwareIds(ids,
+            deleteAllFacetsByMultiFacetAwareIds(dataElementIds,
                                             'delete from datamodel.join_dataelement_to_facet where dataelement_id in :ids')
 
 
-            log.trace('Removing {} DataElements', ids.size())
+            log.trace('Removing {} DataElements', dataElementIds.size())
             Utils.executeInBatches(dataElementIds, {ids ->
                 sessionFactory.currentSession
                     .createSQLQuery('DELETE FROM datamodel.data_element WHERE id IN :ids')

--- a/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataElementService.groovy
+++ b/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataElementService.groovy
@@ -116,20 +116,18 @@ class DataElementService extends ModelItemService<DataElement> implements Summar
         }.id().list() as List<UUID>
 
         if (dataElementIds) {
-            List<List<UUID>> batches = Utils.partition(dataElementIds, BATCH_SIZE)
-            batches.each { ids ->
+            log.trace('Removing facets for {} DataElements', ids.size())
+            deleteAllFacetsByMultiFacetAwareIds(ids,
+                                            'delete from datamodel.join_dataelement_to_facet where dataelement_id in :ids')
 
-                log.trace('Removing facets for {} DataElements', ids.size())
 
-                deleteAllFacetsByMultiFacetAwareIds(ids,
-                                                    'delete from datamodel.join_dataelement_to_facet where dataelement_id in :ids')
-
-                log.trace('Removing {} DataElements', ids.size())
+            log.trace('Removing {} DataElements', ids.size())
+            Utils.executeInBatches(dataElementIds, {ids ->
                 sessionFactory.currentSession
                     .createSQLQuery('DELETE FROM datamodel.data_element WHERE id IN :ids')
                     .setParameter('ids', ids)
                     .executeUpdate()
-            }
+            })
         }
         log.trace('DataElements removed')
     }

--- a/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataElementService.groovy
+++ b/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataElementService.groovy
@@ -116,7 +116,7 @@ class DataElementService extends ModelItemService<DataElement> implements Summar
         }.id().list() as List<UUID>
 
         if (dataElementIds) {
-            log.trace('Removing facets for {} DataElements', ids.size())
+            log.trace('Removing facets for {} DataElements', dataElementIds.size())
             deleteAllFacetsByMultiFacetAwareIds(dataElementIds,
                                             'delete from datamodel.join_dataelement_to_facet where dataelement_id in :ids')
 

--- a/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/datatype/DataTypeService.groovy
+++ b/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/datatype/DataTypeService.groovy
@@ -133,10 +133,13 @@ class DataTypeService extends ModelItemService<DataType> implements DefaultDataT
 
             log.trace('Removing {} DataTypes', dataTypeIds.size())
 
-            sessionFactory.currentSession
-                .createSQLQuery('DELETE FROM datamodel.data_type WHERE data_model_id IN :ids')
-                .setParameter('ids', dataModelIds)
-                .executeUpdate()
+            Utils.executeInBatches(dataModelIds as List, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('DELETE FROM datamodel.data_type WHERE data_model_id IN :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+
+            })
 
             log.trace('DataTypes removed')
         }

--- a/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/datatype/ReferenceTypeService.groovy
+++ b/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/datatype/ReferenceTypeService.groovy
@@ -98,10 +98,12 @@ class ReferenceTypeService extends ModelItemService<ReferenceType> implements Su
                                                 'delete from datamodel.join_datatype_to_facet where datatype_id in :ids')
 
             log.trace('Removing {} ReferenceTypes', referenceTypeIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('DELETE FROM datamodel.data_type WHERE id IN :ids')
-                .setParameter('ids', referenceTypeIds)
-                .executeUpdate()
+            Utils.executeInBatches(referenceTypeIds, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('DELETE FROM datamodel.data_type WHERE id IN :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+            })
 
             log.trace('ReferenceTypes removed')
         }

--- a/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/datatype/enumeration/EnumerationValueService.groovy
+++ b/mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/datatype/enumeration/EnumerationValueService.groovy
@@ -91,11 +91,13 @@ class EnumerationValueService extends ModelItemService<EnumerationValue> impleme
                                                 'delete from datamodel.join_enumerationvalue_to_facet where enumerationvalue_id in :ids')
 
             log.trace('Removing {} EnumerationValues', enumerationValueIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('DELETE FROM datamodel.enumeration_value WHERE id IN :ids')
-                .setParameter('ids', enumerationValueIds)
-                .executeUpdate()
 
+            Utils.executeInBatches(enumerationValueIds, {ids ->
+                sessionFactory.currentSession
+                        .createSQLQuery('DELETE FROM datamodel.enumeration_value WHERE id IN :ids')
+                        .setParameter('ids', ids)
+                        .executeUpdate()
+            })
             log.trace('EnumerationValues removed')
         }
     }

--- a/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/CodeSetService.groovy
+++ b/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/CodeSetService.groovy
@@ -144,26 +144,28 @@ class CodeSetService extends ModelService<CodeSet> {
         GormUtils.disableDatabaseConstraints(sessionFactory as SessionFactoryImplementor)
 
         log.trace('Removing Term relationships to {} CodeSets', idsToDelete.size())
-        sessionFactory.currentSession
-            .createSQLQuery('DELETE FROM terminology.join_codeset_to_term WHERE codeset_id IN :ids')
-            .setParameter('ids', idsToDelete)
-            .executeUpdate()
+        Utils.executeInBatches(idsToDelete as List, {ids ->
+            sessionFactory.currentSession
+                .createSQLQuery('DELETE FROM terminology.join_codeset_to_term WHERE codeset_id IN :ids')
+                .setParameter('ids', ids)
+                .executeUpdate()
 
-        log.trace('Removing facets')
-        deleteAllFacetsByMultiFacetAwareIds(idsToDelete.toList(), 'delete from terminology.join_codeset_to_facet where codeset_id in :ids')
+            log.trace('Removing facets')
+            deleteAllFacetsByMultiFacetAwareIds(ids as List, 'delete from terminology.join_codeset_to_facet where codeset_id in :ids')
 
-        log.trace('Content removed')
-        sessionFactory.currentSession
-            .createSQLQuery('DELETE FROM terminology.code_set WHERE id IN :ids')
-            .setParameter('ids', idsToDelete)
-            .executeUpdate()
+            log.trace('Content removed')
+            sessionFactory.currentSession
+                .createSQLQuery('DELETE FROM terminology.code_set WHERE id IN :ids')
+                .setParameter('ids', ids)
+                .executeUpdate()
 
-        log.trace('CodeSets removed')
+            log.trace('CodeSets removed')
 
-        sessionFactory.currentSession
-            .createSQLQuery('DELETE FROM core.breadcrumb_tree WHERE domain_id IN :ids')
-            .setParameter('ids', idsToDelete)
-            .executeUpdate()
+            sessionFactory.currentSession
+                .createSQLQuery('DELETE FROM core.breadcrumb_tree WHERE domain_id IN :ids')
+                .setParameter('ids', ids)
+                .executeUpdate()
+        })
 
         log.trace('Breadcrumb trees removed')
 

--- a/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/TerminologyService.groovy
+++ b/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/TerminologyService.groovy
@@ -357,10 +357,12 @@ class TerminologyService extends ModelService<Terminology> {
         deleteAllFacetsByMultiFacetAwareIds(idsToDelete.toList(), 'delete from terminology.join_terminology_to_facet where terminology_id in :ids')
 
         log.trace('Content removed')
-        sessionFactory.currentSession
-            .createSQLQuery('DELETE FROM terminology.terminology WHERE id IN :ids')
-            .setParameter('ids', idsToDelete)
-            .executeUpdate()
+        Utils.executeInBatches(idsToDelete as List, {ids ->
+            sessionFactory.currentSession
+                .createSQLQuery('DELETE FROM terminology.terminology WHERE id IN :ids')
+                .setParameter('ids', ids)
+                .executeUpdate()
+        })
 
         log.trace('Terminologies removed')
 

--- a/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/item/TermRelationshipTypeService.groovy
+++ b/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/item/TermRelationshipTypeService.groovy
@@ -141,10 +141,12 @@ class TermRelationshipTypeService extends ModelItemService<TermRelationshipType>
                                                 'delete from terminology.join_termrelationshiptype_to_facet where termrelationshiptype_id in :ids')
 
             log.trace('Removing {} TermRelationshipTypes', termRelationshipTypeIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('DELETE FROM terminology.term_relationship_type WHERE terminology_id IN :ids')
-                .setParameter('ids', modelIds)
-                .executeUpdate()
+            Utils.executeInBatches(modelIds as List, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('DELETE FROM terminology.term_relationship_type WHERE terminology_id IN :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+            })
 
             log.trace('TermRelationshipTypes removed')
 

--- a/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/item/TermService.groovy
+++ b/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/item/TermService.groovy
@@ -102,28 +102,26 @@ class TermService extends ModelItemService<Term> {
         List<UUID> termIds = Term.byTerminologyIdInList(modelIds).id().list() as List<UUID>
 
         if (termIds) {
-            List<List<UUID>> batches = Utils.partition(termIds, BATCH_SIZE)
-            batches.each { batchedIds ->
-
-                log.trace('Removing TermRelationships in {} Terms', batchedIds.size())
+            Utils.executeInBatches(modelIds as List, {ids ->
+                log.trace('Removing TermRelationships in {} Terms', ids.size())
                 termRelationshipService.deleteAllByModelIds(modelIds)
-
-                log.trace('Removing CodeSet references for {} Terms', batchedIds.size())
+                log.trace('Removing CodeSet references for {} Terms', ids.size())
                 sessionFactory.currentSession
                         .createSQLQuery('DELETE FROM terminology.join_codeset_to_term WHERE term_id IN :ids')
-                        .setParameter('ids', batchedIds)
+                        .setParameter('ids', ids)
                         .executeUpdate()
 
-                log.trace('Removing facets for {} Terms', batchedIds.size())
-                deleteAllFacetsByMultiFacetAwareIds(batchedIds,
+                log.trace('Removing facets for {} Terms', ids.size())
+                deleteAllFacetsByMultiFacetAwareIds(ids,
                         'delete from terminology.join_term_to_facet where term_id in :ids')
 
-            }
-            log.trace('Removing {} Terms', termIds.size())
-            sessionFactory.currentSession
-                    .createSQLQuery('DELETE FROM terminology.term WHERE terminology_id IN :ids')
-                    .setParameter('ids', modelIds)
-                    .executeUpdate()
+
+                log.trace('Removing {} Terms', termIds.size())
+                sessionFactory.currentSession
+                        .createSQLQuery('DELETE FROM terminology.term WHERE terminology_id IN :ids')
+                        .setParameter('ids', modelIds)
+                        .executeUpdate()
+            })
 
             log.trace('Terms removed')
         }

--- a/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/item/term/TermRelationshipService.groovy
+++ b/mdm-plugin-terminology/grails-app/services/uk/ac/ox/softeng/maurodatamapper/terminology/item/term/TermRelationshipService.groovy
@@ -96,10 +96,12 @@ class TermRelationshipService extends ModelItemService<TermRelationship> {
                                                 'delete from terminology.join_term_to_facet where term_id in :ids')
 
             log.trace('Removing {} TermRelationships', termRelationshipIds.size())
-            sessionFactory.currentSession
-                .createSQLQuery('DELETE FROM terminology.term_relationship WHERE id IN :ids')
-                .setParameter('ids', termRelationshipIds)
-                .executeUpdate()
+            Utils.executeInBatches(termRelationshipIds, {ids ->
+                sessionFactory.currentSession
+                    .createSQLQuery('DELETE FROM terminology.term_relationship WHERE id IN :ids')
+                    .setParameter('ids', ids)
+                    .executeUpdate()
+            })
 
             log.trace('TermRelationships removed')
         }

